### PR TITLE
Add z=-1 soil layer

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary
Features "Replace z=-1 stone layer with a layer of soil"

#### Purpose of change
https://github.com/CleverRaven/Cataclysm-DDA/issues/48792
This doesn't resolve that issue, but it adds the soil and updates all vanilla mapgen (AFAIK) to have soil instead of rock at z=-1.

#### Describe the solution
`Allow specifying default terrain per z-level` - b6d7018
Instead of just specifying the default overmap terrain at z=0, allow specifying the default overmap terrain at all z levels. Add soil terrain and a solid_earth overmap tile to fill the z=-1 layer.

`Fix hardcoded overmap terrain generation` - edb6c20
Various hardcoded features (ant hills, etc) needed updating for z=-1 not being solid rock.

`Replace rock with earth on z-1 JSON mapgens` - e31653e
Go through a bunch of mapgen and replace the rock floors with concrete floors, and the stone walls with dirt ones.
@Night-Pryanik I'd appreciate your review of the silos and caves.

`Move hardcoded slimepit generation down 1 z level` - a10d05c
Adjust the slimepit overmap special to be a proper pit, so that the hardcoded generation occurs in a rock layer. I didn't want to copy the hardcoded mapgen used there.

`Adjust overmap terrains for z-1 soil level` - 5028a31
Various overmap terrains were made to look like empty rock - adjust the ones on z=-1 to look to solid earth.

`Update hardcoded mapgen for soil on z-1` - c95ba60
Adjust the hardcoded mapgens on z=-1 to use soil and dirt instead of rock and rock floors.

#### Describe alternatives you've considered
Doing more in this PR, particularly modded mapgen. This is quite big as is, and I'd rather it not get bigger.

#### Testing
Generate a new world, run around under the ground.
<details>
<summary>Lots of Pictures</summary>

image

image
image
image
image
image
image
image
image
image
image
</details>

#### Additional context
Damn GitHub forgetting titles when you reload the page!

### CODE REVIEW

**Comment by I-am-Erk** [comment]
Not in this PR, but we should next add some map extras to the soil layer for chunks of rock.

**Comment by I-am-Erk** [comment]
this is awesome, no idea how to interpret this:
```
/home/runner/work/Cataclysm-DDA/Cataclysm-DDA/src/mapgen_functions.cpp:2314:5: error: Static declaration of 't_soil' is dangerous because 'ter_id' is a specialization of int_id and it will not update automatically when game data changes.  Consider switching to a string_id. [cata-static-int_id-constants,-warnings-as-errors]
```

**Comment by actual-nh** [comment]
>```
> /home/runner/work/Cataclysm-DDA/Cataclysm-DDA/src/mapgen_functions.cpp:2314:5: error: Static declaration of 't_soil' is dangerous because 'ter_id' is a specialization of int_id and it will not update automatically when game data changes.  Consider switching to a string_id. [cata-static-int_id-constants,-warnings-as-errors]
> ```

ter_str_id instead of ter_id?

**Comment by anothersimulacrum** [comment]
Yep.

**Comment by chaohedgehogs** [comment]
It might be better if the LMOE and bunkers were moved a z-level down to the new rock layer, with the dirt layer having an enclosed stairwell, given the fact that they're meant to be secure, durable locations.

**Comment by anothersimulacrum** [comment]
Are these sorts of things normally built into the bedrock, or into the soil? I am not sure.

**Comment by chaohedgehogs** [comment]
The entrance to the LMOE looks like a rocky outcrop, so the bedrock would extend to the surface; the LMOE shelter itself could either be on the soil layer but still surrounded by rock, or a few tiles of rock surround a stairwell and the shelter is further down. If the LMOE extends to the edge of the overmap tile then the first option would have the problem of needing more than one overmap tile, which seems undesirable. Given that the newer LMOE variants are professionally constructed, and all are intended for surviving a major war/disaster by a serious prepper it seems likely that they would have been dug into the bedrock layer.

The survival bunker looks like an it should be on the soil layer though, lined with wall supports to prevent cave-ins; the irregular layout and wooden walls imply that it was improvised.

The military bunker would certainly be dug into the bedrock, given the basement is often a secure storage location guarded by turrets and electronic locks.

There's also the possible issue of sounds travelling between z-levels; anything happening in the military bunker or LMOE is unlikely to be heard outside if the entrace is closed, so having an extra z-level separating them avoids that.

**Comment by anothersimulacrum** [comment]
> The entrance to the LMOE looks like a rocky outcrop, so the bedrock would extend to the surface; the LMOE shelter itself could either be on the soil layer but still surrounded by rock, or a few tiles of rock surround a stairwell and the shelter is further down.

I changed it to soil as well.
I'll make changes as you suggest.

**Comment by I-am-Erk** [comment]
Really you wouldn't dig an lmoe into the living rock. You'd pour concrete into a pit. The one homemade one in a rocky cave could be an exception

File: data/json/furniture_and_terrain/terrain-zlevel-transitions.json [hunk 1]
```diff
@@ -209,6 +209,48 @@
...
+    "description": "The upper end of an earth ramp leading up.",
+    "symbol": "<",
+    "color": "brown",
+    "move_cost": 2,
+    "flags": [ "RAMP_UP" ]
+  },
+  {
+    "type": "terrain",
+    "id": "t_earth_ramp_up_low",
+    "name": "road ramp up (low end)",
```
Comment (cainiaowu) [inline data/json/furniture_and_terrain/terrain-zlevel-transitions.json:247]: Do you mean` earth ramp up (low end)`？

**Comment by Night-Pryanik** [comment]
I didn't make it in time to review the changes.

File: data/json/furniture_and_terrain/terrain-zlevel-transitions.json (hunk 1 above)
Comment (anothersimulacrum) [reply data/json/furniture_and_terrain/terrain-zlevel-transitions.json:247]: Yes.

File: data/json/mapgen/cave.json [hunk 2]
```diff
@@ -34,9 +34,9 @@
       ],
       "terrain": {
         ".": [ [ "t_region_groundcover_forest", 4 ], [ "t_region_tree", 1 ] ],
-        "|": "t_rock",
-        ",": "t_rock_roof",
-        "%": [ "t_rock", "t_region_groundcover_forest" ],
+        "|": "t_soil",
+        ",": "t_dirt_underground",
+        "%": [ "t_soil", "t_region_groundcover_forest" ],
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:39]: I think this is wrong. The "walls" of a cave shouldn't consist of soil. This OMT is an enter to rocky cave above ground, not its underground part.

File: data/json/mapgen/cave.json [hunk 3]
```diff
@@ -85,10 +85,10 @@
       ],
       "terrain": {
         ".": [ [ "t_region_groundcover_forest", 4 ], [ "t_region_tree", 1 ] ],
-        "|": "t_rock",
-        "%": [ "t_rock", "t_region_groundcover_forest" ],
+        "|": "t_soil",
+        "%": [ "t_soil", "t_region_groundcover_forest" ],
         ";": "t_dirt",
-        ",": "t_rock_roof",
+        ",": "t_dirt_underground",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:91]: The same here.

File: data/json/mapgen/cave.json [hunk 4]
```diff
@@ -129,11 +129,11 @@
...
         ".": [ [ "t_region_groundcover_forest", 4 ], [ "t_region_tree", 1 ] ],
-        "|": "t_rock",
-        "%": [ "t_rock", "t_region_groundcover_forest" ],
+        "|": "t_soil",
+        "%": [ "t_soil", "t_region_groundcover_forest" ],
         ";": "t_dirt",
-        ",": "t_rock_roof",
-        "W": "t_rock_roof",
+        ",": "t_dirt_underground",
+        "W": "t_dirt_underground",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:136]: The same here.

File: data/json/mapgen/cave.json [hunk 5]
```diff
@@ -176,11 +176,11 @@
...
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:183]: The same here.

File: data/json/mapgen/cave.json [hunk 6]
```diff
@@ -222,11 +222,11 @@
...
         ".": [ [ "t_region_groundcover_forest", 4 ], [ "t_region_tree", 1 ] ],
-        "|": "t_rock",
-        "%": [ "t_rock", "t_region_groundcover_forest" ],
+        "|": "t_soil",
+        "%": [ "t_soil", "t_region_groundcover_forest" ],
         ";": "t_dirt",
-        ",": "t_rock_roof",
-        "B": "t_rock_roof",
+        ",": "t_dirt_underground",
+        "B": "t_dirt_underground",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:229]: The same here.

File: data/json/mapgen/cave.json [hunk 7]
```diff
@@ -268,12 +268,12 @@
       ],
       "terrain": {
         ".": [ [ "t_region_groundcover_forest", 4 ], [ "t_region_tree", 1 ] ],
-        "|": "t_rock",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:271]: The same here.

File: data/json/mapgen/cave.json [hunk 8]
```diff
@@ -316,12 +316,12 @@
...
-        "%": [ "t_rock", "t_region_groundcover_forest" ],
+        "|": "t_soil",
+        "%": [ "t_soil", "t_region_groundcover_forest" ],
         ";": "t_dirt",
-        ",": "t_rock_roof",
-        "B": "t_rock_roof",
-        "W": "t_rock_roof",
+        ",": "t_dirt_underground",
+        "B": "t_dirt_underground",
+        "W": "t_dirt_underground",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:324]: The same here.

File: data/json/mapgen/cave.json [hunk 9]
```diff
@@ -363,12 +363,12 @@
...
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:371]: The same here.

File: data/json/mapgen/cave.json [hunk 10]
```diff
@@ -411,17 +411,17 @@
...
+        "%": [ "t_soil", "t_region_groundcover_forest" ],
+        ";": "t_dirt",
+        "!": "t_dirt",
+        "?": "t_dirt",
+        "#": "t_dirt",
+        "~": "t_dirt",
+        "@": "t_dirt_underground",
+        ",": "t_dirt_underground",
+        "q": "t_dirt_underground",
+        "R": "t_dirt"
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:424]: The same here.

File: data/json/mapgen/cave.json [hunk 11]
```diff
@@ -479,7 +479,7 @@
         "............,...........",
         "............>..........."
       ],
-      "terrain": { ".": "t_rock", ",": "t_rock_floor", "W": "t_rock_floor", ">": "t_slope_up", "~": "t_water_dp" },
+      "terrain": { ".": "t_soil", ",": "t_dirt", "W": "t_dirt", ">": "t_slope_up", "~": "t_water_dp" },
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:482]: `t_dirt` here should be changed to `t_dirt_underground` as it's an underground section of the cave with a "roof" above.

File: data/json/mapgen/cave.json [hunk 12]
```diff
@@ -497,7 +497,7 @@
     "om_terrain": [ "cave_underground" ],
     "weight": 1000,
     "object": {
-      "fill_ter": "t_rock_floor",
+      "fill_ter": "t_dirt",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:500]: The same here.

File: data/json/mapgen/cave.json [hunk 13]
```diff
@@ -525,7 +525,7 @@
         "............,...........",
         "............>..........."
       ],
-      "terrain": { ".": "t_rock", ",": "t_rock_floor", ">": "t_slope_up" },
+      "terrain": { ".": "t_soil", ",": "t_dirt", ">": "t_slope_up" },
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:528]: The same here.

File: data/json/mapgen/cave.json [hunk 14]
```diff
@@ -564,7 +564,7 @@
         "............ ...........",
         "............>..........."
       ],
-      "terrain": { ".": "t_rock", " ": "t_rock_floor", ">": "t_slope_up" },
+      "terrain": { ".": "t_soil", " ": "t_dirt", ">": "t_slope_up" },
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:567]: The same here.

File: data/json/mapgen/cave.json [hunk 15]
```diff
@@ -575,7 +575,7 @@
     "om_terrain": [ "cave_underground" ],
     "weight": 500,
     "object": {
-      "fill_ter": "t_rock_floor",
+      "fill_ter": "t_dirt",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:578]: The same here.

File: data/json/mapgen/cave.json [hunk 16]
```diff
@@ -603,7 +603,7 @@
         "...........,,,..........",
         "............>..........."
       ],
-      "terrain": { ".": "t_rock", ",": "t_rock_floor", "!": "t_rock_floor", ">": "t_slope_up" },
+      "terrain": { ".": "t_soil", ",": "t_dirt", "!": "t_dirt", ">": "t_slope_up" },
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:606]: The same here.

File: data/json/mapgen/cave.json [hunk 17]
```diff
@@ -618,7 +618,7 @@
     "om_terrain": [ "cave_underground" ],
     "weight": 250,
     "object": {
-      "fill_ter": "t_rock_floor",
+      "fill_ter": "t_dirt",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:621]: The same here.

File: data/json/mapgen/cave.json [hunk 18]
```diff
@@ -646,7 +646,7 @@
         "............ ....F.7J G.",
         "............>..........."
       ],
-      "terrain": { ".": "t_rock", " ": "t_rock_floor", ">": "t_slope_up" },
+      "terrain": { ".": "t_soil", " ": "t_dirt", ">": "t_slope_up" },
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:649]: The same here.

File: data/json/mapgen/cave.json [hunk 19]
```diff
@@ -688,7 +688,7 @@
     "om_terrain": [ "cave_underground" ],
     "weight": 700,
     "object": {
-      "fill_ter": "t_rock_floor",
+      "fill_ter": "t_dirt",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:691]: The same here.

File: data/json/mapgen/cave.json [hunk 20]
```diff
@@ -716,7 +716,7 @@
         "........... 2 ..........",
         "............>..........."
       ],
-      "terrain": { ".": "t_rock", " ": "t_rock_floor", ">": "t_slope_up", "2": "t_railroad_track_small" },
+      "terrain": { ".": "t_soil", " ": "t_dirt", ">": "t_slope_up", "2": "t_railroad_track_small" },
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:719]: The same here.

File: data/json/mapgen/cave.json [hunk 21]
```diff
@@ -728,7 +728,7 @@
     "method": "json",
     "om_terrain": [ "cave_rat_underground" ],
     "object": {
-      "fill_ter": "t_rock_floor",
+      "fill_ter": "t_dirt",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:731]: The same here.

File: data/json/mapgen/cave.json [hunk 22]
```diff
@@ -756,7 +756,7 @@
         "......................  ",
         "...................... <"
       ],
-      "terrain": { ".": "t_rock", " ": "t_rock_floor", ">": "t_slope_up", "<": "t_slope_down" },
+      "terrain": { ".": "t_soil", " ": "t_dirt", ">": "t_slope_up", "<": "t_slope_down" },
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:759]: The same here.

File: data/json/mapgen/cave.json [hunk 23]
```diff
@@ -771,7 +771,7 @@
     "method": "json",
     "om_terrain": [ "cave_rat" ],
     "object": {
-      "fill_ter": "t_rock_floor",
+      "fill_ter": "t_dirt",
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:774]: The same here.

File: data/json/mapgen/cave.json [hunk 24]
```diff
@@ -800,10 +800,10 @@
...
       ],
       "terrain": {
-        " ": "t_rock_floor",
-        ".": "t_rock",
-        "|": [ [ "t_rock", 20 ], [ "t_rock_floor", 80 ] ],
-        "%": [ "t_rock", "t_rock_floor" ],
+        " ": "t_dirt",
+        ".": "t_soil",
+        "|": [ [ "t_soil", 20 ], [ "t_dirt", 80 ] ],
+        "%": [ "t_soil", "t_dirt" ],
```
Comment (Night-Pryanik) [inline data/json/mapgen/cave.json:806]: The same for `t_dirt` here.

File: data/json/mapgen/lmoe.json [hunk 25]
```diff
@@ -117,7 +117,7 @@
     "//2": "This particular 4-bed shelter is definitely homemade, mostly carved out of bare rock, and seems to be incomplete.",
     "weight": 100,
     "object": {
-      "fill_ter": "t_rock_floor",
+      "fill_ter": "t_dirt",
```
Comment (Night-Pryanik) [inline data/json/mapgen/lmoe.json:120]: Probably the same here.

**Pull request merged**